File: go/nvmagic/guardrails/guardrails.go
// SPDX-FileCopyrightText: Copyright (c) 2026, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

// Package guardrails provides shorthand access to NVMagic guardrail registration.
//
// Guardrails are priority-ordered middleware that sanitize or gate tool and LLM
// calls. They run in priority order (lower values first). Function names drop
// the "Guardrail" suffix found in the parent nvmagic package.
//
// Three guardrail categories are supported for both tools and LLMs:
//   - SanitizeRequest: modifies outgoing request arguments/parameters.
//   - SanitizeResponse: modifies incoming response data.
//   - ConditionalExecution: gates whether the call should proceed at all.
//
// Example usage:
//
//	import "github.com/nvidia/nvmagic/go/nvmagic/guardrails"
//
//	// Register a tool request sanitizer that redacts sensitive fields.
//	err := guardrails.RegisterToolSanitizeRequest("redact-pii", 10,
//	    func(name string, args json.RawMessage) json.RawMessage {
//	        // ... redact PII from args ...
//	        return args
//	    },
//	)
//
//	// Later, remove it.
//	_ = guardrails.DeregisterToolSanitizeRequest("redact-pii")
package guardrails

import (
	"encoding/json"

	"github.com/nvidia/nvmagic/go/nvmagic"
)

// --- Tool Sanitize Request ---

// RegisterToolSanitizeRequest registers a guardrail that sanitizes tool request
// arguments before they are passed to the tool. The callback receives the tool
// name and arguments JSON and must return the (possibly modified) arguments.
// Guardrails run in priority order (lower values first). This is a shorthand
// for [nvmagic.RegisterToolSanitizeRequestGuardrail].
func RegisterToolSanitizeRequest(name string, priority int32, fn nvmagic.ToolSanitizeFunc) error {
	return nvmagic.RegisterToolSanitizeRequestGuardrail(name, priority, fn)
}

// DeregisterToolSanitizeRequest removes a tool sanitize-request guardrail by
// name. This is a shorthand for [nvmagic.DeregisterToolSanitizeRequestGuardrail].
func DeregisterToolSanitizeRequest(name string) error {
	return nvmagic.DeregisterToolSanitizeRequestGuardrail(name)
}

// --- Tool Sanitize Response ---

// RegisterToolSanitizeResponse registers a guardrail that sanitizes tool
// response data before it is returned to the caller. The callback receives the
// tool name and response JSON and must return the (possibly modified) response.
// This is a shorthand for [nvmagic.RegisterToolSanitizeResponseGuardrail].
func RegisterToolSanitizeResponse(name string, priority int32, fn nvmagic.ToolSanitizeFunc) error {
	return nvmagic.RegisterToolSanitizeResponseGuardrail(name, priority, fn)
}

// DeregisterToolSanitizeResponse removes a tool sanitize-response guardrail by
// name. This is a shorthand for [nvmagic.DeregisterToolSanitizeResponseGuardrail].
func DeregisterToolSanitizeResponse(name string) error {
	return nvmagic.DeregisterToolSanitizeResponseGuardrail(name)
}

// --- Tool Conditional Execution ---

// RegisterToolConditionalExecution registers a guardrail that conditionally
// gates tool execution. The callback returns nil to allow execution or a
// non-nil pointer to an error message string to reject it. This is a shorthand
// for [nvmagic.RegisterToolConditionalExecutionGuardrail].
func RegisterToolConditionalExecution(name string, priority int32, fn nvmagic.ToolConditionalFunc) error {
	return nvmagic.RegisterToolConditionalExecutionGuardrail(name, priority, fn)
}

// DeregisterToolConditionalExecution removes a tool conditional-execution
// guardrail by name. This is a shorthand for
// [nvmagic.DeregisterToolConditionalExecutionGuardrail].
func DeregisterToolConditionalExecution(name string) error {
	return nvmagic.DeregisterToolConditionalExecutionGuardrail(name)
}

// --- LLM Sanitize Request ---

// RegisterLlmSanitizeRequest registers a guardrail that sanitizes the LLM
// request JSON before the call is made. This is a shorthand for
// [nvmagic.RegisterLlmSanitizeRequestGuardrail].
func RegisterLlmSanitizeRequest(name string, priority int32, fn nvmagic.JSONFunc) error {
	return nvmagic.RegisterLlmSanitizeRequestGuardrail(name, priority, fn)
}

// DeregisterLlmSanitizeRequest removes an LLM sanitize-request guardrail by
// name. This is a shorthand for [nvmagic.DeregisterLlmSanitizeRequestGuardrail].
func DeregisterLlmSanitizeRequest(name string) error {
	return nvmagic.DeregisterLlmSanitizeRequestGuardrail(name)
}

// --- LLM Sanitize Response ---

// RegisterLlmSanitizeResponse registers a guardrail that sanitizes LLM response
// data before it is returned to the caller. The callback receives serialized
// LLMResponse JSON (containing a "data" field). This is a shorthand for
// [nvmagic.RegisterLlmSanitizeResponseGuardrail].
func RegisterLlmSanitizeResponse(name string, priority int32, fn nvmagic.LLMResponseFunc) error {
	return nvmagic.RegisterLlmSanitizeResponseGuardrail(name, priority, fn)
}

// DeregisterLlmSanitizeResponse removes an LLM sanitize-response guardrail by
// name. This is a shorthand for [nvmagic.DeregisterLlmSanitizeResponseGuardrail].
func DeregisterLlmSanitizeResponse(name string) error {
	return nvmagic.DeregisterLlmSanitizeResponseGuardrail(name)
}

// --- LLM Conditional Execution ---

// RegisterLlmConditionalExecution registers a guardrail that conditionally
// gates LLM execution. The callback receives LLM request parameters and returns
// nil to allow execution or a non-nil pointer to an error message string to
// reject it. This is a shorthand for
// [nvmagic.RegisterLlmConditionalExecutionGuardrail].
func RegisterLlmConditionalExecution(name string, priority int32, fn nvmagic.LLMConditionalFunc) error {
	return nvmagic.RegisterLlmConditionalExecutionGuardrail(name, priority, fn)
}

// DeregisterLlmConditionalExecution removes an LLM conditional-execution
// guardrail by name. This is a shorthand for
// [nvmagic.DeregisterLlmConditionalExecutionGuardrail].
func DeregisterLlmConditionalExecution(name string) error {
	return nvmagic.DeregisterLlmConditionalExecutionGuardrail(name)
}

// --- Tool Conditional Execution (standalone) ---

// ToolConditionalExecution runs the registered tool conditional execution
// guardrail chain. Returns nil if all pass, or an error if blocked. This is a
// shorthand for [nvmagic.ToolConditionalExecution].
func ToolConditionalExecution(name string, args json.RawMessage) error {
	return nvmagic.ToolConditionalExecution(name, args)
}

// --- LLM Conditional Execution (standalone) ---

// LlmConditionalExecution runs the registered LLM conditional execution
// guardrail chain. Returns nil if all pass, or an error if blocked. This is a
// shorthand for [nvmagic.LlmConditionalExecution].
func LlmConditionalExecution(request json.RawMessage) error {
	return nvmagic.LlmConditionalExecution(request)
}

File: python/nat_nexus/guardrails.py
# SPDX-FileCopyrightText: Copyright (c) 2026, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0

"""Guardrail registration for tools and LLMs.

Guardrails run inside the middleware pipeline and can sanitize or gate requests
and responses. They are priority-ordered (ascending) and registered by name.

Sanitize guardrails are currently observability-oriented in the managed
``execute`` APIs: they affect the payload captured on lifecycle events, but
they do not rewrite the values passed to the user callable or returned to the
caller.

**Tool guardrails** — callback signatures:

    register_tool_sanitize_request(name, priority, fn)
        ``fn(tool_name: str, args: Any) -> Any`` — sanitize tool arguments.

    register_tool_sanitize_response(name, priority, fn)
        ``fn(tool_name: str, result: Any) -> Any`` — sanitize tool result.

    register_tool_conditional_execution(name, priority, fn)
        ``fn(tool_name: str, args: Any) -> Optional[str]`` — return ``None``
        to allow, or a rejection reason to block.

**LLM guardrails** — callback signatures:

    register_llm_sanitize_request(name, priority, fn)
        ``fn(request: LLMRequest) -> LLMRequest`` — sanitize the LLM request.

    register_llm_sanitize_response(name, priority, fn)
        ``fn(response: dict) -> dict`` — sanitize the LLM response.

    register_llm_conditional_execution(name, priority, fn)
        ``fn(request: LLMRequest) -> Optional[str]`` — return ``None``
        to allow, or a rejection reason to block.

Each ``register_*`` has a corresponding ``deregister_*`` that takes the name
and returns ``True`` if a guardrail was found and removed.

Example::

    import nat_nexus

    def redact_pii(tool_name, args):
        # Sanitize PII from tool arguments
        return {k: "***" if "ssn" in k else v for k, v in args.items()}

    nat_nexus.guardrails.register_tool_sanitize_request("pii-filter", 10, redact_pii)
"""

from typing import Any, Callable, Optional

from nat_nexus._native import LLMRequest
from nat_nexus._native import (
    nat_nexus_deregister_llm_conditional_execution_guardrail as _native_deregister_llm_conditional_execution,
)
from nat_nexus._native import (
    nat_nexus_deregister_llm_sanitize_request_guardrail as _native_deregister_llm_sanitize_request,
)
from nat_nexus._native import (
    nat_nexus_deregister_llm_sanitize_response_guardrail as _native_deregister_llm_sanitize_response,
)
from nat_nexus._native import (
    nat_nexus_deregister_tool_conditional_execution_guardrail as _native_deregister_tool_conditional_execution,
)
from nat_nexus._native import (
    nat_nexus_deregister_tool_sanitize_request_guardrail as _native_deregister_tool_sanitize_request,
)
from nat_nexus._native import (
    nat_nexus_deregister_tool_sanitize_response_guardrail as _native_deregister_tool_sanitize_response,
)
from nat_nexus._native import (
    nat_nexus_register_llm_conditional_execution_guardrail as _native_register_llm_conditional_execution,
)
from nat_nexus._native import (
    nat_nexus_register_llm_sanitize_request_guardrail as _native_register_llm_sanitize_request,
)
from nat_nexus._native import (
    nat_nexus_register_llm_sanitize_response_guardrail as _native_register_llm_sanitize_response,
)
from nat_nexus._native import (
    nat_nexus_register_tool_conditional_execution_guardrail as _native_register_tool_conditional_execution,
)
from nat_nexus._native import (
    nat_nexus_register_tool_sanitize_request_guardrail as _native_register_tool_sanitize_request,
)
from nat_nexus._native import (
    nat_nexus_register_tool_sanitize_response_guardrail as _native_register_tool_sanitize_response,
)

# ---------------------------------------------------------------------------
# Tool guardrails
# ---------------------------------------------------------------------------

ToolSanitizeGuardrail = Callable[[str, Any], Any]
ToolConditionalExecutionGuardrail = Callable[[str, Any], Optional[str]]
LlmSanitizeRequestGuardrail = Callable[[LLMRequest], LLMRequest]
LlmSanitizeResponseGuardrail = Callable[[dict], dict]
LlmConditionalExecutionGuardrail = Callable[[LLMRequest], Optional[str]]


def register_tool_sanitize_request(name: str, priority: int, guardrail: ToolSanitizeGuardrail) -> None:
    """Register a tool sanitize-request guardrail.

    The guardrail callback receives the tool name and arguments and returns
    sanitized arguments for the emitted ``Start`` event payload.

    Args:
        name: Unique guardrail name.
        priority: Priority (ascending order; lower runs first).
        guardrail: Callable ``(tool_name: str, args: Any) -> Any``.

    Raises:
        RuntimeError: If a guardrail with this name already exists.
    """
    return _native_register_tool_sanitize_request(name, priority, guardrail)


def deregister_tool_sanitize_request(name: str) -> bool:
    """Remove a tool sanitize-request guardrail.

    Args:
        name: Name of the guardrail to remove.

    Returns:
        ``True`` if a guardrail with the given name was found and removed,
        ``False`` otherwise.
    """
    return _native_deregister_tool_sanitize_request(name)


def register_tool_sanitize_response(name: str, priority: int, guardrail: ToolSanitizeGuardrail) -> None:
    """Register a tool sanitize-response guardrail.

    The guardrail callback receives the tool name and result and returns
    a sanitized result for the emitted ``End`` event payload.

    Args:
        name: Unique guardrail name.
        priority: Priority (ascending order; lower runs first).
        guardrail: Callable ``(tool_name: str, result: Any) -> Any``.

    Raises:
        RuntimeError: If a guardrail with this name already exists.
    """
    return _native_register_tool_sanitize_response(name, priority, guardrail)


def deregister_tool_sanitize_response(name: str) -> bool:
    """Remove a tool sanitize-response guardrail.

    Args:
        name: Name of the guardrail to remove.

    Returns:
        ``True`` if a guardrail with the given name was found and removed,
        ``False`` otherwise.
    """
    return _native_deregister_tool_sanitize_response(name)


def register_tool_conditional_execution(name: str, priority: int, guardrail: ToolConditionalExecutionGuardrail) -> None:
    """Register a tool conditional-execution guardrail.

    The guardrail callback receives the tool name and arguments and returns
    ``None`` to allow execution or a rejection reason string to block it.

    Args:
        name: Unique guardrail name.
        priority: Priority (ascending order; lower runs first).
        guardrail: Callable ``(tool_name: str, args: Any) -> Optional[str]``.

    Raises:
        RuntimeError: If a guardrail with this name already exists.
    """
    return _native_register_tool_conditional_execution(name, priority, guardrail)


def deregister_tool_conditional_execution(name: str) -> bool:
    """Remove a tool conditional-execution guardrail.

    Args:
        name: Name of the guardrail to remove.

    Returns:
        ``True`` if a guardrail with the given name was found and removed,
        ``False`` otherwise.
    """
    return _native_deregister_tool_conditional_execution(name)


# ---------------------------------------------------------------------------
# LLM guardrails
# ---------------------------------------------------------------------------


def register_llm_sanitize_request(name: str, priority: int, guardrail: LlmSanitizeRequestGuardrail) -> None:
    """Register an LLM sanitize-request guardrail.

    The guardrail callback receives the LLM request and returns a sanitized
    ``LLMRequest`` for the emitted ``Start`` event payload.

    Args:
        name: Unique guardrail name.
        priority: Priority (ascending order; lower runs first).
        guardrail: Callable ``(request: LLMRequest) -> LLMRequest``.

    Raises:
        RuntimeError: If a guardrail with this name already exists.
    """
    return _native_register_llm_sanitize_request(name, priority, guardrail)


def deregister_llm_sanitize_request(name: str) -> bool:
    """Remove an LLM sanitize-request guardrail.

    Args:
        name: Name of the guardrail to remove.

    Returns:
        ``True`` if a guardrail with the given name was found and removed,
        ``False`` otherwise.
    """
    return _native_deregister_llm_sanitize_request(name)


def register_llm_sanitize_response(name: str, priority: int, guardrail: LlmSanitizeResponseGuardrail) -> None:
    """Register an LLM sanitize-response guardrail.

    The guardrail callback receives the LLM response dict and returns a
    sanitized dict for the emitted ``End`` event payload.

    Args:
        name: Unique guardrail name.
        priority: Priority (ascending order; lower runs first).
        guardrail: Callable ``(response: dict) -> dict``.

    Raises:
        RuntimeError: If a guardrail with this name already exists.
    """
    return _native_register_llm_sanitize_response(name, priority, guardrail)


def deregister_llm_sanitize_response(name: str) -> bool:
    """Remove an LLM sanitize-response guardrail.

    Args:
        name: Name of the guardrail to remove.

    Returns:
        ``True`` if a guardrail with the given name was found and removed,
        ``False`` otherwise.
    """
    return _native_deregister_llm_sanitize_response(name)


def register_llm_conditional_execution(name: str, priority: int, guardrail: LlmConditionalExecutionGuardrail) -> None:
    """Register an LLM conditional-execution guardrail.

    The guardrail callback receives the LLM request and returns ``None``
    to allow execution or a rejection reason string to block it.

    Args:
        name: Unique guardrail name.
        priority: Priority (ascending order; lower runs first).
        guardrail: Callable ``(request: LLMRequest) -> Optional[str]``.

    Raises:
        RuntimeError: If a guardrail with this name already exists.
    """
    return _native_register_llm_conditional_execution(name, priority, guardrail)


def deregister_llm_conditional_execution(name: str) -> bool:
    """Remove an LLM conditional-execution guardrail.

    Args:
        name: Name of the guardrail to remove.

    Returns:
        ``True`` if a guardrail with the given name was found and removed,
        ``False`` otherwise.
    """
    return _native_deregister_llm_conditional_execution(name)


__all__ = [
    "register_tool_sanitize_request",
    "deregister_tool_sanitize_request",
    "register_tool_sanitize_response",
    "deregister_tool_sanitize_response",
    "register_tool_conditional_execution",
    "deregister_tool_conditional_execution",
    "register_llm_sanitize_request",
    "deregister_llm_sanitize_request",
    "register_llm_sanitize_response",
    "deregister_llm_sanitize_response",
    "register_llm_conditional_execution",
    "deregister_llm_conditional_execution",
]
